### ISSUE TITLE
Support GeoLite2 format

### ISSUE THREAD

**Issue by okkez** [y-ken/fluent-plugin-geoip#39]
* http://dev.maxmind.com/geoip/geoip2/geolite2/

I've found https://github.com/mosuka/fluent-plugin-filter-geoip
This is fluentd filter plugin that supports GeoLite2 format.

fluent-plugin-geoip does not support GeoLite2 format.

There are some bindings for libmaxmand. http://dev.maxmind.com/geoip/geoip2/downloadable/#MaxMind_APIs
I create benchmark script to compare them: 

https://gist.github.com/okkez/b1cf88a57c9ff168005335c644b0a532

How do you think about supporting GeoLite2 format?

**Comment by y-ken** [comment]
Thank you for compairing the related gems!
It sounds good to support geoLite2 format if get almost same throughput.

**Comment by okkez** [comment]
Thank you for comment!
I will keep investigating this issue.

**Comment by repeatedly** [comment]
geoip2_compat seems to have good performance and API.
How about using geoip2_compat for GeoLite2 support?

**Comment by okkez** [comment]
There are some differences between GeoLite and GeoLite2.
GeoLite has country_code3,dma_code,area_code but GeoLite2 doesn't.

I think we can support GeoLite2 as a isolated new feature.
For now we implement GeoLite support as `Fluent::GeoIP`.
We can implement GeoLite2 support as `Fluent::GeoIp2` and add some new configurations for `Fluent::GeoIP2`.

**Comment by okkez** [comment]
I'm trying to support GoIP2 on following branch:
https://github.com/okkez/fluent-plugin-geoip/tree/support-geoip2

Almost done but not tested.
This branch is already migrated to v0.14 API.

I will test and add test code for this branch in these days.

**Comment by repeatedly** [comment]
Many users can't use v0.14 on production.
So geoip2 works with only v0.14 seems not good for users.

**Comment by okkez** [comment]
Right.
I can backport geoip2 support for v0.12 after test.

**Comment by okkez** [comment]
NOTE: Some distributions don't have libmaxminddb-dev package

- don't have: precise, trusty, jessie
- have: xenial, stretch, sid, centos7(epel), centos6(epel)

**Comment by okkez** [comment]
hive_geoip2 is so slow and is not maintained.
I've released new gem geoip2_c. This is faster than hive_geoip2 always and faster than geoip2_compat in some case.

I've created benchmark script using geoip, geoip2_compat, geoip2_c in fluent-plugin-geoip (filter_geoip).
https://gist.github.com/okkez/4c72d7ce127a89246010cce492b3abe3

**Comment by y-ken** [comment]
sounds nice!!! i felt it is the time to adopt geoip2.

**Comment by okkez** [comment]
@y-ken Thank you for replying.
I'm going to implement geoip2 support.

BTW, do you need td-agent 1 support with geoip2?
I think it is so hard to support td-agent 1.
Ruby1.9.3 has been already EOL and Fluentd v0.10.x will be [EOL soon](https://groups.google.com/forum/#!topic/fluentd/c4f9j9VMWkA).
Because:
* some old distributions does not have libmaxminddb-dev package
  * It is hard to construct CI environment for old Fluentd and old Ruby, I think.
  * I can construct CI environment with [Dcoker](https://hub.docker.com/_/ruby/) for Ruby 2.1 or later, I will use Alpine Linux for CI.
* geoip2_c uses `Hash#dig`, `Array#dig` method internal for performance and simplicity of implementation
  * `#dig` has been added since Ruby2.3, but some gem library adds `#dig` for Ruby2.1 and Ruby2.2, so I cannot support Ruby2.0.0 or earlier.

cc/ @repeatedly

**Comment by gcavalcante8808** [comment]
Is there any update for this Issue?

I don't know if this is directly related, but when I try to use the following record:

```
city            ${city.names.en["user.ip"]}
```

The Fluentd Daemon returns the following error:s returning the error:

```
2017-06-20 19:54:27 +0000 [error]: config error file="/fluentd/etc/fluent.conf" error="geoip: unsupported key city.names.en"
```
I just downloaded the latest version of City Lite Database Format v2 from the maxmind site.

Gems Installed:

bigdecimal (1.2.8)
cool.io (1.5.0)
did_you_mean (1.0.0)
dig_rb (1.0.1)
elasticsearch (1.0.18)
elasticsearch-api (1.0.18)
elasticsearch-transport (1.0.18)
excon (0.57.0)
faraday (0.12.1)
fluent-mixin-rewrite-tag-name (0.1.0)
fluent-plugin-elasticsearch (1.9.2)
fluent-plugin-geoip (0.7.0)
fluentd (0.12.36)
geoip-c (0.9.1)
http_parser.rb (0.6.0)
io-console (0.4.5)
json (2.1.0, 1.8.3)
minitest (5.9.0)
msgpack (1.1.0)
multi_json (1.12.1)
multipart-post (2.0.0)
net-telnet (0.1.1)
oj (3.1.4)
power_assert (0.2.7)
psych (2.1.0)
rake (10.5.0)
rdoc (4.2.1)
sigdump (0.2.4)
string-scrub (0.0.5)
test-unit (3.1.7)
thread_safe (0.3.6)
tzinfo (1.2.3)
tzinfo-data (1.2017.2)
yajl-ruby (1.3.0)

Thanks in advance.

**Comment by okkez** [comment]
@gcavalcante8808 

GeoIP2 support is optional for fluent-plugin-geoip 0.7.0.
You must install geoip2_c or geoip2_compat gem if you use `geoip2_*` backend.
Because we don't add these dependencies to gemspec for now.
See also https://github.com/y-ken/fluent-plugin-geoip#dependency

And then you must specify `backend_library` config param in your fluent.conf.

```
<store>
  @type geoip
  backend_library geoip2_c # or geoip2_compat
  # ... snip
</store>
```

**Comment by okkez** [comment]
Released v1.0.0.
